Zend/zend_operators.c

```diff
@@ -742,6 +754,12 @@ ZEND_API zend_long ZEND_FASTCALL _zval_get_long_func(zval *op) /* {{{ */
 		case IS_DOUBLE:
 			return zend_dval_to_lval(Z_DVAL_P(op));
 		case IS_STRING:
+			if (!silent) {
+				if (is_numeric_string(Z_STRVAL_P(op), Z_STRLEN_P(op), NULL, NULL, -1)==0) {
+					zend_error(E_WARNING, "A non-numeric value encountered");
+					return 0;
+				}
+			}
```

Comment:
This part is troublesome. If we use `is_numeric_string` to convert, then we start supporting scientific notation for numbers, which we don't at present because `strtol()` doesn't support it.

If we don't use it, though, then there's the problem Dmitry points out: `is_numeric_string` will accept things that `strtol` won't. :/